test/integration/replicaset/replicaset_test.go

```diff
@@ -272,246 +280,595 @@ func TestAdoption(t *testing.T) {
+	rss, _ := createRSsPods(t, c, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{}, ns.Name)
+	rs = rss[0]
 
 	stopCh := make(chan struct{})
 	informers.Start(stopCh)
-	waitToObservePods(t, podInformer, 2)
 	go rm.Run(5, stopCh)
-	waitRSStable(t, clientSet, rs, ns.Name)
-
-	// change the rs's selector to match both pods
-	patch := `{"spec":{"selector":{"matchLabels": {"uniqueKey":"1"}},"template":{"metadata":{"labels":{"uniqueKey":"1"}}}}}`
-	rsClient := clientSet.Extensions().ReplicaSets(ns.Name)
-	rs, err := rsClient.Patch(rs.Name, types.StrategicMergePatchType, []byte(patch))
+	defer close(stopCh)
+	waitRSStable(t, c, rs, ns.Name)
+
+	testStatusReplicas(t, c, rs, ns.Name, 3)
+	testStatusReplicas(t, c, rs, ns.Name, 0)
+	testStatusReplicas(t, c, rs, ns.Name, 2)
+
+	// Update 2 pods that RS created such that one is failed and one is being deleted
+	// Ensure both are being created again in the end
+	podClient := c.Core().Pods(ns.Name)
+	podSelector := labels.Set{"foo": "bar"}.AsSelector()
+	options := metav1.ListOptions{LabelSelector: podSelector.String()}
+	pods, err := podClient.List(options)
 	if err != nil {
-		t.Fatalf("Failed to patch replica set: %v", err)
+		t.Fatalf("Failed obtaining a list of pods that match the pod labels %v: %v", labelMap, err)
 	}
-	t.Logf("patched rs = %#v", rs)
-	// wait for the rs to create one more pod
-	if err := wait.Poll(10*time.Second, 60*time.Second, func() (bool, error) {
-		return verifyRemainingObjects(t, clientSet, ns.Name, 1, 3)
+	if pods == nil {
+		t.Fatalf("Obtained a nil list of pods")
+	}
+	if len(pods.Items) != 2 {
+		t.Fatalf("len(pods) = %v, want %v", len(pods.Items), 2)
+	}
+
+	failedPod := &pods.Items[0]
+	deletingPod := &pods.Items[1]
+	if err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		newPod, err := podClient.Get(failedPod.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		newPod.Status.Phase = v1.PodFailed
+		_, err = podClient.Update(newPod)
+		return err
 	}); err != nil {
-		t.Fatal(err)
+		t.Fatalf("Failed to set .Status.Phase of pod %s: %v", failedPod.Name, err)
 	}
-	podClient := clientSet.Core().Pods(ns.Name)
-	pod2, err = podClient.Get(pod2.Name, metav1.GetOptions{})
-	if err != nil {
-		t.Fatalf("Failed to get pod2: %v", err)
+	if err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		newPod, err := podClient.Get(deletingPod.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		return c.Core().Pods(ns.Name).Delete(newPod.Name, &metav1.DeleteOptions{})
+	}); err != nil {
+		t.Fatalf("Failed to delete pod %s: %v", deletingPod.Name, err)
 	}
-	if len(pod2.OwnerReferences) != 0 {
-		t.Fatalf("ownerReferences of pod2 is not cleared, got %#v", pod2.OwnerReferences)
+
+	rsClient := c.Extensions().ReplicaSets(ns.Name)
+	if err := wait.PollImmediate(interval, timeout, func() (bool, error) {
+		newRS, err := rsClient.Get(rs.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		if newRS.Status.Replicas != 2 {
```

Comment:
This is racy. It could be that you get here before the RS has noticed any of the changes you made to the Pods. In that case, Status.Replicas would also be 2 so you can't distinguish whether you raced or whether the test passed.

In this case, you could verify the Pods themselves rather than asking RS what it sees. Poll the list of Pods from the server (don't use informer cache; it might not have seen the changes you made yet), and wait until you see a count of 2 Pods that are neither Failed nor pending deletion. Those 2 Pods must have been created by the RS. Note that there might be extra Pods (e.g. the one pending deletion) that you should ignore.

Reply:
@enisoc You are right. The `Status.Replicas` check may happen before the RS observes the pod changes.
What do you mean when you mentioned `don't use informer cache`?

I am thinking about something like this:
```
podClient := c.Core().Pods(ns.Name)
// poll all pods from the API server
options := metav1.ListOptions{LabelSelector: labels.Everything().String()}
pods, err := podClient.List(options)
```

Reply:
In general use the ObservedGeneration of the RS. Poll until it increments.

Reply:
Sure!

Reply:
@kow3ns In this case, we are changing Pods, not the RS Spec. ObservedGeneration doesn't say anything about whether the RS has observed a given Pod change.

Reply:
Noted, thanks for the clarification!